### PR TITLE
Support removing python from zfs/grub closure

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change
ZFS (and consequently grub) depends on python, I don't want that because it adds ~50M. With this PR you can remove python from its closure with the overlay
```nix
self: super: {
  zfs = super.zfs.override { enablePython = false; };
}
```

This overlay reduces the closure of `grub2` from 249.9M to 190.7M, and the closure of ZFS from 236.0M to 176.7M.

No defaults were changed.

Ping @Mic92 @sorki @FRidh @edolstra @wkennington 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -16918,7 +16918,7 @@ in
 
     zenpower = callPackage ../os-specific/linux/zenpower { };
 
-    inherit (callPackage ../os-specific/linux/zfs {
+    inherit (callPackages ../os-specific/linux/zfs {
```
Comment (jonringer) [inline pkgs/top-level/all-packages.nix:16921]: unrelated to PR, but what's the difference between callPackage and callPackages

I could only find https://github.com/NixOS/nixpkgs/issues/36354 and what's in the source:
```
callPackages = lib.callPackagesWith splicedPackagesWithXorg;
```
but that still doesn't explain what's going on.

Comment (Infinisil) [reply pkgs/top-level/all-packages.nix:16921]: `callPackage <file>` is for when the file looks like
```nix
{ stdenv }: stdenv.mkDerivation { ... }
```

It makes the result of the call contain a `.override` attribute you can use to change the arguments.

`callPackages <file>` is for when the file looks like
```nix
{ stdenv }: {
  foo = stdenv.mkDerivation { ... };
  bar = stdenv.mkDerivation { ... };
}
```

It makes all the _attributes_ of the result contain a `.override` attribute you can use to change the arguments for any package individually.

If `callPackage` is used on the latter, you'd get
```nix
{
  foo = ...;
  bar = ...;
  override = ...;
}
```
as a result, without any `.override` on the actual derivations, which is why you couldn't `.override` on `pkgs.zfs` before this change.

Comment (jonringer) [reply pkgs/top-level/all-packages.nix:16921]: ah, that makes sense. I feel like this should be documented somewhere more formally.

Thank you for the explanation.

**Comment by Mic92** [comment]
Looks fine although I wonder what the usecase is because zfs is usually used on systems with plenty resources (RAM/storage space).

**Comment by Infinisil** [comment]
@Mic92 I'm using this on a system that gets deployed via DigitalOcean images (using https://github.com/NixOS/nixpkgs/pull/66978). The bigger the image, the longer it takes to upload (and possibly costs more and takes more time to load), so I'm trying to keep that to a minimum.

**Comment by sorki** [comment]
+1 to minification. Regarding small systems ZFS is now enabled even on Raspberry Pi which doesn't seem very useful.

**Review by jonringer — APPROVED** [review]
LGTM

failures are broken on master
```
https://github.com/NixOS/nixpkgs/pull/84129
6 package marked as broken and skipped:
linuxPackages_hardkernel_4_14.zfs linuxPackages_hardkernel_4_14.zfsStable linuxPackages_hardkernel_4_14.zfsUnstable linuxPackages_hardkernel_latest.zfs linuxPackages_hardkernel_latest.zfsStable linuxPackages_hardkernel_latest.zfsUnstable

7 package failed to build:
linuxPackages_5_6.zfs linuxPackages_latest-libre.zfs linuxPackages_latest_hardened.zfs linuxPackages_latest_xen_dom0.zfs linuxPackages_latest_xen_dom0_hardened.zfs linuxPackages_testing_bcachefs.zfs linuxPackages_testing_hardened.zfs

10 package built:
linuxPackages-libre.zfs linuxPackages.zfs linuxPackages_4_14.zfs linuxPackages_4_19.zfs linuxPackages_4_4.zfs linuxPackages_4_9.zfs linuxPackages_5_5.zfs linuxPackages_hardened.zfs linuxPackages_xen_dom0.zfs linuxPackages_xen_dom0_hardened.zfs
```

**Comment by Infinisil** [comment]
Hm wait, why does it rebuild stuff anyways, I don't think this should be

**Comment by Infinisil** [comment]
Indeed, I had a mistake in there! I fixed it now, so there shouldn't be any rebuilds anymore.

**Review by jonringer — APPROVED** [review]
LGTM

```
Nothing changed
https://github.com/NixOS/nixpkgs/pull/84129
$ nix-shell /home/jon/.cache/nixpkgs-review/pr-84129-1/shell.nix
```

**Pull request merged**